src/symbolic/simplify_program.jl

```diff
@@ -111,24 +111,52 @@ function get_program_rules(alg, shash)
         (@rule sieve(true, ~a) => a),
         (@rule sieve(false, ~a) => block()), #TODO should add back skipvisitor
 
+
+        (@rule loop(~idx, ~ext::isvirtual, ~body) => begin
+            #body_contain_idx = mapreduce((node) -> node.kind === index && node == idx, 
+            #                             (x, y) -> x || y, PostOrderDFS(body), init=false) 
+            body_contain_idx = idx ∈ getunbound(body)
+            if !body_contain_idx
+                decl_in_scope = Set()
+                ScopeVisitor(scope=decl_in_scope)(body) 
```

Comment:
I didn't realize ScopeVisitor does this. One could also do:

```
decl_in_scope = filter(!isnothing, map(node-> if @capture(node, declare(~tns, ~init)) tns end, PostOrderDFS(body)))
```

Reply:
Yeah, using ScopeVisitor is excessive for this problem. I replaced with your suggestion.